extensions/parquet/table/src/main/java/io/deephaven/parquet/table/util/S3SeekableChannelProvider.java

```diff
@@ -0,0 +1,124 @@
+import software.amazon.awssdk.services.s3.model.HeadObjectResponse;
+import software.amazon.awssdk.http.crt.AwsCrtAsyncHttpClient;
+
+import java.io.IOException;
+import java.net.URI;
+import java.nio.channels.SeekableByteChannel;
+import java.nio.file.Path;
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeoutException;
+
+import static io.deephaven.parquet.table.util.S3SeekableByteChannel.CONNECTION_TIMEOUT_MINUTES;
+import static io.deephaven.parquet.table.util.S3SeekableByteChannel.CONNECTION_TIMEOUT_UNIT;
+
+/**
+ * {@link SeekableChannelsProvider} implementation that is used to fetch objects from AWS S3 instances.
+ */
+public final class S3SeekableChannelProvider implements SeekableChannelsProvider {
+
+    private final S3AsyncClient s3AsyncClient;
+
+    /**
+     * Parquet file {@link URI} stored as a {@link Path} to save on parsing time at the time of constructing new read
+     * channels. Note that this conversion is lossy, as the {@link URI} will contain characters "s3://" but {@link Path}
+     * will have "s3:/".
+     */
+    private final Path parquetFilePath;
+    private final String bucket, key;
+    private final long size;
+    private final int fragmentSize;
+    private final int maxCacheSize;
+    private final int readAheadCount;
+    private final Map<Long, ChannelContext> contextMap = new HashMap<>(); // TODO Remove this
+
+
+    public S3SeekableChannelProvider(final URI parquetFileURI, final ParquetInstructions readInstructions) {
+        if (!(readInstructions.getSpecialInstructions() instanceof S3ParquetInstructions)) {
+            throw new IllegalArgumentException("Must provide S3ParquetInstructions to read files from S3");
+        }
+        final S3ParquetInstructions s3Instructions = (S3ParquetInstructions) readInstructions.getSpecialInstructions();
+        final String awsRegionName = s3Instructions.awsRegionName();
+        final int maxConcurrentRequests = s3Instructions.maxConcurrentRequests();
+        final SdkAsyncHttpClient asyncHttpClient = AwsCrtAsyncHttpClient.builder()
+                .maxConcurrency(maxConcurrentRequests)
+                .connectionTimeout(Duration.ofMinutes(CONNECTION_TIMEOUT_MINUTES))
+                .build();
+        this.s3AsyncClient = S3AsyncClient.builder()
+                .region(Region.of(awsRegionName))
+                .httpClient(asyncHttpClient)
+                .build();
+
+        this.fragmentSize = s3Instructions.fragmentSize();
+        this.maxCacheSize = s3Instructions.maxCacheSize();
+        this.readAheadCount = s3Instructions.readAheadCount();
+
+        this.parquetFilePath = Path.of(parquetFileURI.toString());
+        this.bucket = parquetFileURI.getHost();
+        this.key = parquetFileURI.getPath().substring(1);
+        // Send a blocking HEAD request to S3 to get the size of the file
```

Comment:
I kept it here because all the channels will need this size of file.
I can send a request on first read but then each thread would need to fetch it at least once.
I can store the size in the context object so that a single thread doesn't need to fetch it again.